src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala

```diff
@@ -900,6 +887,30 @@ abstract class BTypes {
 
   final case class MethodBType(argumentTypes: List[BType], returnType: BType) extends BType
 
+  object BTypeExporter {
+    private[this] val builderTL: ThreadLocal[StringBuilder] = new ThreadLocal[StringBuilder](){
+      override protected val initialValue: StringBuilder = new StringBuilder(64)
+    }
+
+    final def btypeToString(btype: BType): String = {
+      val builder = builderTL.get()
```

Comment:
It isn't ideal to use a `ThreadLocal` without ever calling `remove()` from each `Thread` that touches it. `ThreadLocal` is implemented with a hash map stored in each `Thread`.

The implementation contains this warning:

```
     * ThreadLocalMap is a customized hash map suitable only for
     * maintaining thread local values. No operations are exported
     * outside of the ThreadLocal class. The class is package private to
     * allow declaration of fields in class Thread.  To help deal with
     * very large and long-lived usages, the hash table entries use
     * WeakReferences for keys. However, since reference queues are not
     * used, stale entries are guaranteed to be removed only when
     * the table starts running out of space.
```

I've seen performance effects like this show up in compiler benchmarking before.

Organizing the right place to call `remove()` though can be fairly tricky with the backend parallelism implementation.

I was thinking of adding a wrapper to allow conveniently removing this thread local from all relevant threads, but the `ThreadLocal` API doesn't allow it.

```scala

import java.io.Closeable

import scala.collection.mutable.ArrayBuffer

final class CloseableThreadLocal[T](initialValue: => T) extends Closeable {
  private val threadLocal = new ThreadLocal[T] {
    override def initialValue(): T = CloseableThreadLocal.this.initialValue
  }
  private val threads = ArrayBuffer[Thread]()
  def get: T = {
    rememberThread()
    threadLocal.get()
  }
  def set(t: T): Unit = {
    rememberThread()
    threadLocal.set(t)
  }
  override def close(): Unit = {
    threads.synchronized {
      threads.foreach { thread =>
        val map = threadLocal.getMap(thread) // getMap is package private, we can't call this :(
        if (map != null) map.remove(threadLocal)
      }
    }
  }

  private def rememberThread(): Unit = {
    threads.synchronized {
      threads += Thread.currentThread()
    }
  }
}
```